### PR TITLE
Pass pre-adjusted cairo context to wxGraphicsContext::CreateFromNative()

### PR DESCRIPTION
**Pull request by AliKet**
It took me awhile to figure out that `wxGTKCairoDCImpl::SetLayoutDirection()` has no effect, but to set the `m_layoutDir` only!
and that we should pass to `wxGraphicsContext::CreateFromNative()` a pre adjusted cairo context for things to work properly.

you can test these samples (with master):
 - drawing
 - griddemo
 - splitter

try some interactions with them and see what happens!

**drawing sample for ex.:**
- the canvas keeps flipping whenever you click inside!
- the rubber-band is not working correctly!

@paulcor thanks in advance for your help on this.

### CODE REVIEW

**Comment by AliKet** [comment]
I use this for testing:
```
diff --git a/samples/aui/auidemo.cpp b/samples/aui/auidemo.cpp
index 666e8dbc2e..aa68b5ee8f 100644
--- a/samples/aui/auidemo.cpp
+++ b/samples/aui/auidemo.cpp
@@ -40,6 +40,9 @@ class MyApp : public wxApp
 {
 public:
     bool OnInit() wxOVERRIDE;
+
+    virtual wxLayoutDirection GetLayoutDirection() const wxOVERRIDE
+        { return wxLayout_RightToLeft; }
 };
 
 wxDECLARE_APP(MyApp);
diff --git a/samples/drawing/drawing.cpp b/samples/drawing/drawing.cpp
index 354271f13b..2b49e900eb 100644
--- a/samples/drawing/drawing.cpp
+++ b/samples/drawing/drawing.cpp
@@ -89,6 +89,9 @@ public:
 
     virtual int OnExit() wxOVERRIDE { DeleteBitmaps(); return 0; }
 
+    virtual wxLayoutDirection GetLayoutDirection() const wxOVERRIDE
+        { return wxLayout_RightToLeft; }
+
 protected:
     void DeleteBitmaps();
 
diff --git a/samples/grid/griddemo.h b/samples/grid/griddemo.h
index 0957d40b8d..25d6022d5c 100644
--- a/samples/grid/griddemo.h
+++ b/samples/grid/griddemo.h
@@ -17,6 +17,9 @@ class GridApp : public wxApp
 {
 public:
     bool OnInit() wxOVERRIDE;
+
+    virtual wxLayoutDirection GetLayoutDirection() const wxOVERRIDE
+        { return wxLayout_RightToLeft; }
 };
 
 
diff --git a/samples/listctrl/listtest.h b/samples/listctrl/listtest.h
index c40ee918ca..4d0dfc2915 100644
--- a/samples/listctrl/listtest.h
+++ b/samples/listctrl/listtest.h
@@ -24,6 +24,9 @@ public:
 
     virtual bool OnInit() wxOVERRIDE;
 
+    virtual wxLayoutDirection GetLayoutDirection() const wxOVERRIDE
+        { return wxLayout_RightToLeft; }
+
 private:
     wxDECLARE_NO_COPY_CLASS(MyApp);
 };
diff --git a/samples/sashtest/sashtest.h b/samples/sashtest/sashtest.h
index af940c2a56..c8976fca8f 100644
--- a/samples/sashtest/sashtest.h
+++ b/samples/sashtest/sashtest.h
@@ -15,6 +15,9 @@ class MyApp: public wxApp
 {
   public:
     bool OnInit(void) wxOVERRIDE;
+
+    virtual wxLayoutDirection GetLayoutDirection() const wxOVERRIDE
+        { return wxLayout_RightToLeft; }
 };
 
 class MyCanvas: public wxScrolledWindow
diff --git a/samples/splitter/splitter.cpp b/samples/splitter/splitter.cpp
index 99c1f930e0..6d29d94ddb 100644
--- a/samples/splitter/splitter.cpp
+++ b/samples/splitter/splitter.cpp
@@ -73,6 +73,9 @@ public:
 
     virtual bool OnInit() wxOVERRIDE;
 
+    virtual wxLayoutDirection GetLayoutDirection() const wxOVERRIDE
+        { return wxLayout_RightToLeft; }
+
     wxDECLARE_NO_COPY_CLASS(MyApp);
 };
 
diff --git a/samples/treectrl/treetest.h b/samples/treectrl/treetest.h
index 43439ea8d5..84210b7c8d 100644
--- a/samples/treectrl/treetest.h
+++ b/samples/treectrl/treetest.h
@@ -26,6 +26,9 @@ public:
 
     bool OnInit() wxOVERRIDE;
 
+    virtual wxLayoutDirection GetLayoutDirection() const wxOVERRIDE
+        { return wxLayout_RightToLeft; }
+
     void SetShowImages(bool show) { m_showImages = show; }
     bool ShowImages() const { return m_showImages; }
 
diff --git a/samples/treelist/treelist.cpp b/samples/treelist/treelist.cpp
index af6905cecb..3271b8caa7 100644
--- a/samples/treelist/treelist.cpp
+++ b/samples/treelist/treelist.cpp
@@ -162,6 +162,9 @@ class MyApp : public wxApp
 {
 public:
     virtual bool OnInit() wxOVERRIDE;
+
+    virtual wxLayoutDirection GetLayoutDirection() const wxOVERRIDE
+        { return wxLayout_RightToLeft; }
 };
 
 // ----------------------------------------------------------------------------
diff --git a/samples/widgets/widgets.cpp b/samples/widgets/widgets.cpp
index 5d83b2461e..b136b0d0ca 100644
--- a/samples/widgets/widgets.cpp
+++ b/samples/widgets/widgets.cpp
@@ -151,6 +151,9 @@ public:
     // return: if OnInit() returns false, the application terminates)
     virtual bool OnInit() wxOVERRIDE;
 
+    virtual wxLayoutDirection GetLayoutDirection() const wxOVERRIDE
+        { return wxLayout_RightToLeft; }
+
     // real implementation of WidgetsPage method with the same name
     bool IsUsingLogWindow() const;
```

File: include/wx/gtk/dc.h [hunk 1]
```diff
@@ -42,6 +42,9 @@ class wxGTKCairoDCImpl: public wxGCDCImpl
     // Set m_size from the given (valid) GdkWindow.
     void InitSize(GdkWindow* window);
 
+    // Adjust cairo context for RTL.
+    void AdjustCairoContext(cairo_t* cr);
```
Comment (vadz) [inline include/wx/gtk/dc.h:46]: If it's only going to be used for RTL, I'd prefer to use a more explicit name for it, e.g. `ApplyLayoutDirection()`.

**Review by vadz — APPROVED** [review]
Thanks, this looks good to me, but I'd prefer if @paulcor could review this too before merging.

P.S. The CI failure is unrelated.

**Comment by paulcor** [comment]
Thanks, I missed that changes to the Cairo context made after `wxGraphicsContext::CreateFromNative()` will be overwritten. Note that there is no need to do anything for `wxScreenDC`, it can't be used for drawing with GTK3. Closed via 62c119fa1efd8c9603890bcf121095150fe96181

**Comment by AliKet** [comment]
> Thanks, I missed that changes to the Cairo context made after `wxGraphicsContext::CreateFromNative()` will be overwritten. Note that there is no need to do anything for `wxScreenDC`, it can't be used for drawing with GTK3. Closed via [62c119f](https://github.com/wxWidgets/wxWidgets/commit/62c119fa1efd8c9603890bcf121095150fe96181)

I didn't know that `wxScreenDC` doesn't work with GTK3, thanks for the info.

**Comment by AliKet** [comment]
`wxClientDCImpl::wxClientDCImpl` is missing a call to `SetLayoutDirection(wxLayout_Default);`

@paulcor

**Pull request closed without merging**